test/helpers/kubectl.go

```diff
@@ -0,0 +1,311 @@
+// Copyright 2018 Authors of Cilium
+// Modified for integration with kube-router
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helpers
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/cloudnativelabs/kube-router/test/ginkgo-ext"
+	. "github.com/onsi/gomega"
+	"github.com/sirupsen/logrus"
+)
+
+// GetCurrentK8SEnv returns the value of K8S_VERSION from the OS environment.
+func GetCurrentK8SEnv() string { return os.Getenv("K8S_VERSION") }
+
+// Kubectl is a wrapper around an SSHMeta. It is used to run Kubernetes-specific
+// commands on the node which is accessible via the SSH metadata stored in its
+// SSHMeta.
+type Kubectl struct {
+	*SSHMeta
+	// *serviceCache
+}
+
+// CreateKubectl initializes a Kubectl helper with the provided vmName and log
+// It marks the test as Fail if cannot get the ssh meta information or cannot
+// execute a `ls` on the virtual machine.
+func CreateKubectl(vmName string, log *logrus.Entry) *Kubectl {
+	node := GetVagrantSSHMeta(vmName)
+	if node == nil {
+		ginkgoext.Fail(fmt.Sprintf("Cannot connect to vmName  '%s'", vmName), 1)
+		return nil
+	}
+	// This `ls` command is a sanity check, sometimes the meta ssh info is not
+	// nil but new commands cannot be executed using SSH, tests failed and it
+	// was hard to debug.
+	res := node.Exec("ls /tmp/")
+	if !res.WasSuccessful() {
+		ginkgoext.Fail(fmt.Sprintf(
+			"Cannot execute ls command on vmName '%s'", vmName), 1)
+		return nil
+	}
+	node.logger = log
+
+	return &Kubectl{
+		SSHMeta: node,
+	}
+}
+
+// ExpectAllPodsTerminated is a wrapper around helpers/WaitCleanAllTerminatingPods.
+// It asserts that the error returned by that function is nil.
+func (kub *Kubectl) ExpectAllPodsTerminated() {
+	err := kub.WaitCleanAllTerminatingPods(HelperTimeout)
+	ExpectWithOffset(1, err).To(BeNil(), "terminating containers are not deleted after timeout")
+}
+
+// WaitCleanAllTerminatingPods waits until all nodes that are in `Terminating`
+// state are deleted correctly in the platform. In case of excedding the
+// given timeout (in seconds) it returns an error
+func (kub *Kubectl) WaitCleanAllTerminatingPods(timeout int64) error {
+	body := func() bool {
+		res := kub.Exec(fmt.Sprintf(
```

Comment:
Any reason for not using client go here? That would give us more flexibility on getting/creating/deleting pods and would be a bit cleaner IMO

Reply:
Guessing this is just from copying it over from Cilium?

Reply:
Yeah, I pretty much stuck to using the cilium approach. Being able to issues other commands beyond kubectl is also a nice benefit of this approach. We could always add "k8s.io/client-go" in the future if we have the need to do more complex stuff...